src/tactic/simps.lean

```diff
@@ -18,29 +18,66 @@ open tactic expr
   If `add_simp` then we make the resulting lemma a simp-lemma. -/
-meta def simps_add_projection (nm : name) (type lhs rhs : expr) (args : list expr)
-  (univs : list name) (add_simp : bool) : tactic unit := do
-  eq_ap ← mk_mapp `eq $ [type, lhs, rhs].map some,
-  refl_ap ← mk_app `eq.refl [type, lhs],
-  decl_name ← get_unused_decl_name nm,
-  let decl_type := eq_ap.pis args,
-  let decl_value := refl_ap.lambdas args,
-  let decl := declaration.thm decl_name univs decl_type (pure decl_value),
-  add_decl decl <|> fail format!"failed to add projection lemma {decl_name}.",
-  when add_simp $
-    set_basic_attribute `simp decl_name tt >> set_basic_attribute `_refl_lemma decl_name tt
+meta def simps_add_projection (nm : name) (type lhs rhs pr : expr) (args : list expr)
+  (univs : list name) (add_simp : bool) : tactic unit :=
+do eq_ap ← mk_mapp `eq [type, lhs, rhs],
+   refl_ap ← mk_app `eq.refl [type, lhs],
+   decl_name ← get_unused_decl_name nm,
+   let decl_type := eq_ap.pis args,
+   let decl_value := refl_ap.lambdas args,
+   let decl := declaration.thm decl_name univs decl_type (pure decl_value),
+   add_decl decl <|> fail format!"failed to add projection lemma {decl_name}.",
+   when add_simp $
+     set_basic_attribute `simp decl_name tt >> set_basic_attribute `_refl_lemma decl_name tt
+
+meta def simps_add_projection' (nm : name) (type lhs rhs : expr) (pr : option expr) (args : list expr)
+  (univs : list name) (add_simp : bool) : tactic unit :=
+do decl_value ← match pr with
+                | none :=
+                  do refl_ap ← mk_app `eq.refl [type, lhs],
+                     pure $ refl_ap.lambdas args
+                | (some pr) := pure pr
+                end,
+   simps_add_projection nm type lhs rhs decl_value args univs add_simp
+
+open native
+
+@[user_attribute]
+meta def projection_attr : user_attribute (rb_lmap name (name × name)) name :=
+{ name := `projection,
+  descr := "descr",
+  cache_cfg := { mk_cache := λ ls,
+                 do { ls.mfoldl (λ m l,
+                       do d ← projection_attr.get_param l,
+                          decl ← get_decl l,
+                          (_,t) ← mk_local_pis decl.type,
+                          (lhs,rhs) ← match_eq t,
+                          pure $ m.insert d (lhs.get_app_fn.const_name,l) ) (rb_lmap.mk _ _) },
+                 dependencies := [] },
+  parser := lean.parser.ident }
+
+meta def get_ext_projections (env : environment) (n : name) : tactic (list $ name × name) :=
+do m ← projection_attr.get_cache,
+   pure $ m.find n
+
+meta def apply_congr : list expr → expr → tactic expr
+| [] p := pure p
+| (v :: vs) p :=
+  mk_congr_fun p v >>= apply_congr vs
 
 /-- Derive lemmas specifying the projections of the declaration.
   If `todo` is non-empty, it will generate exactly the names in `todo`. -/
 meta def simps_add_projections : ∀(e : environment) (nm : name) (suffix : string)
```

Comment:
The current logic in this function is a little weird: it uses the argument `pr` if it's at a "leaf" (i.e. when it calls `simps_add_projection'`), but it forgets about it in the recursive calls. This will lead to problems if we nest custom projections: for example if we also define a custom projection for `pprod` in your test below.